src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -214,24 +232,39 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                             _document is Document &&
                             !await TryAddCachedDocumentDiagnosticsAsync(stateSet, AnalysisKind.Semantic, list, cancellationToken).ConfigureAwait(false))
                         {
-                            // Check whether we want up-to-date document wide semantic diagnostics
-                            var spanBased = analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis();
-                            if (!_blockForData && !spanBased)
+                            if (!_incrementalAnalysis)
                             {
-                                containsFullResult = false;
+                                if (!_blockForData)
+                                {
+                                    containsFullResult = false;
+                                }
+                                else
+                                {
+                                    semanticSpanBasedAnalyzers.Add(stateSet);
+                                }
                             }
                             else
                             {
-                                var stateSets = spanBased ? semanticSpanBasedAnalyzers : semanticDocumentBasedAnalyzers;
-                                stateSets.Add(stateSet);
+                                // We can perform incremental analysis only for analyzers that support
+                                // span-based semantic diagnostic analysis.
+                                var spanBased = analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis();
+                                if (!_blockForData && !spanBased)
+                                {
+                                    containsFullResult = false;
```

Comment:
something is nagging me here with all this logic that makes me feel like i want the _blockForData pulled out higher.... but maybe that's not possible?  

It just seems that the all the branches of all these if/elses are subtly close to each other.   

Would it instead make sense to have a helper that gives you the stateSets value to add to?  If it returns a value, you add stateSet to it.  If it doesn't, you set containsFullResult to false. a nd you do all this :)

Reply:
> something is nagging me here with all this logic that makes me feel like i want the _blockForData pulled out higher.... but maybe that's not possible?

That is correct. `_blockForData` is essentially determining whether to use only cached diagnostics, or force computation in case these are not already computed and cached. We can't move this up because this is the core place where we attempt to look into the diagnostics cache and make this decision.

For additional context, this flag is passed as true from the code path that wants to show a lightbulb vs screwdriver in the editor gutter. It is basically a polling approach to check if there are any definite code fixes for diagnostics on the line, and if so show a lightbulb instead of a screwdriver. We don't intend to force compute any diagnostics on this path as there is no real urgency in when the lightbulb/screwdriver shows up in the gutter, we are okay to wait for background analysis to compute and cache full document diagnostics for the active document before it to show up.

Reply:
> Would it instead make sense to have a helper that gives you the stateSets value to add to? If it returns a value, you add stateSet to it. If it doesn't, you set containsFullResult to false. a nd you do all this :)

That sounds good. I'll make that change.